### PR TITLE
[SPARK-49307][CONNECT][SQL] Add Kryo serialization to agnostic encoder framework.

### PR DESCRIPTION
**Pull request by hvanhovell**
### What changes were proposed in this pull request?
This PR adds support for Kryo based serialization using the AgnosticEncoder framework. Please note that Kryo serialization is only support for Classic.

### Why are the changes needed?
This PR is a stepping stone for centralizing the Encoders object in sql/api.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Added tests.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

**Review by HyukjinKwon — APPROVED** [review]

**Comment by dongjoon-hyun** [comment]
Could you re-trigger the failed CI, @hvanhovell ?
```
[info] *** 1 TEST FAILED ***
[error] Failed: Total 1339, Failed 1, Errors 0, Passed 1338, Ignored 5
[error] Failed tests:
[error] 	org.apache.spark.sql.connect.client.arrow.ArrowEncoderSuite
```

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/arrow/ArrowEncoderSuite.scala [hunk 1]
```diff
@@ -776,6 +776,16 @@ class ArrowEncoderSuite extends ConnectFunSuite with BeforeAndAfterAll {
     }
   }
 
+  test("kryo serialization") {
```
Comment (dongjoon-hyun) [inline connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/arrow/ArrowEncoderSuite.scala:779]: Oh, this newly added test case failed. Could you take a look at the failure?

```
[info] - kryo serialization *** FAILED *** (3 milliseconds)
[info]   Expected exception org.apache.spark.SparkException to be thrown, but org.apache.spark.SparkRuntimeException was thrown (ArrowEncoderSuite.scala:780)
```

Comment (hvanhovell) [reply connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/arrow/ArrowEncoderSuite.scala:779]: yeah definitely my bad. Will update tonight.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. (Pending CIs).

**Comment by dongjoon-hyun** [comment]
Merged to master. Thank you, @hvanhovell and @HyukjinKwon .

**Pull request closed without merging**